### PR TITLE
TW-514 Use 3route API for exchange rates calculation

### PR DESCRIPTION
**Pull request by keshan3262**

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
 export const ALICE_BOB_PUBLIC_KEY = getEnv('ALICE_BOB_PUBLIC_KEY');
+export const THREE_ROUTE_API_URL = getEnv('THREE_ROUTE_API_URL');
+export const THREE_ROUTE_API_AUTH_TOKEN = getEnv('THREE_ROUTE_API_AUTH_TOKEN');
 
-if (!Boolean(QUIPUSWAP_FA12_FACTORIES)) throw new Error('process.env.QUIPUSWAP_FA12_FACTORIES not found.');
-if (!Boolean(QUIPUSWAP_FA2_FACTORIES)) throw new Error('process.env.QUIPUSWAP_FA2_FACTORIES not found.');
-if (!Boolean(MOONPAY_SECRET_KEY)) throw new Error('process.env.MOONPAY_SECRET_KEY not found.');
-if (!Boolean(ALICE_BOB_PRIVATE_KEY)) throw new Error('process.env.ALICE_BOB_PRIVATE_KEY not found.');
-if (!Boolean(ALICE_BOB_PUBLIC_KEY)) throw new Error('process.env.ALICE_BOB_PUBLIC_KEY not found.');
+const variablesToAssert = [
```
[the remark this answers is not in the document]
Comment (keshan3262) [reply src/config.ts:14]: No, they are different. For example, for `MOONPAY_SECRET_KEY=foo` `name` is `'MOONPAY_SECRET_KEY'` but `value` is `'foo'`

File: src/utils/tzkt.ts [hunk 2]
```diff
@@ -99,58 +92,59 @@ export type TzktTokenData = {
...
+        } | null;
+      }[]
+    | null;
+}
 
-export const tokensMetadataProvider = new DataProvider(
-  24 * 3600 * 1000,
-  (network: TZKT_NETWORKS, address?: string, token_id?: number) => {
-    const getTokensMetadata =
-      network === TZKT_NETWORKS.MAINNET
```
[the remark this answers is not in the document]
Comment (keshan3262) [reply src/utils/tzkt.ts:116]: As far as I know, this service is used only for mainnet. @unixvb please confirm or deny

File: src/config.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (keshan3262) [reply src/config.ts:13]: Already done, see https://github.com/madfish-solutions/templewallet-backend/pull/102/files#diff-c3095d5010e65c52737a98a5d618ea24049ebe90c8470752426081d70ed6e012R22

File: src/utils/makeBuildQueryFn.ts [hunk 3]
```diff
@@ -1,27 +1,15 @@
...
-
-function isAbsoluteURL(url) {
-  // A URL is considered absolute if it begins with "<scheme>://" or "//" (protocol-relative URL).
-  // RFC 3986 defines scheme name as a sequence of characters beginning with a letter and followed
-  // by any combination of letters, digits, plus, period, or hyphen.
-  return /^([a-z][a-z\d\+\-\.]*:)?\/\//i.test(url);
-}
-
-export default function makeBuildQueryFn<P, R>(baseUrl: string, maxConcurrentQueries?: number) {
+export default function makeBuildQueryFn<P extends object, R>(
```
Comment (herkoss) [inline src/utils/makeBuildQueryFn.ts:8]:
```suggestion
export function makeBuildQueryFn<P extends object, R>(
```

Comment (herkoss) [reply src/utils/makeBuildQueryFn.ts:8]: Let's use named exports

**Review by herkoss — APPROVED** [review]
GJ!

File: src/utils/makeBuildQueryFn.ts (hunk 3 above)
Comment (keshan3262) [reply src/utils/makeBuildQueryFn.ts:8]: It isn't a tradition in this project but OK

**Review by NovakVlad2424 — APPROVED** [review]
done

**Pull request merged**